src/sleep/sleep.c

```diff
@@ -263,40 +263,78 @@ static int execute(
 }
 
 static int execute_s2h(const SleepConfig *sleep_config) {
-        _cleanup_close_ int tfd = -1;
-        struct itimerspec ts = {};
         int r;
-
         assert(sleep_config);
+        usec_t suspend_interval = sleep_config->hibernate_delay_sec, before_timestamp, after_timestamp;
 
-        tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
-        if (tfd < 0)
-                return log_error_errno(errno, "Error creating timerfd: %m");
+        do {
+                _cleanup_close_ int tfd = -1;
+                struct itimerspec ts = {};
+                int last_capacity, current_capacity, previous_discharge_rate, estimated_discharge_rate = 0;
+                double time_elapsed;
 
-        log_debug("Set timerfd wake alarm for %s",
-                  FORMAT_TIMESPAN(sleep_config->hibernate_delay_sec, USEC_PER_SEC));
+                tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
+                if (tfd < 0)
+                        return log_error_errno(errno, "Error creating timerfd: %m");
 
-        timespec_store(&ts.it_value, sleep_config->hibernate_delay_sec);
+                /* Calculate current battery capacity and current time before suspension*/
+                last_capacity = read_battery_capacity_percentage();
+                if (last_capacity < 0)
+                        return log_error_errno(errno, "Error fetching battery capacity percentage: %m");
```

Comment:
`errno` shouldn't be used here as it isn't set by `read_battery_capacity_percentage()`